src/python/grpcio_tests/tests/unit/_compression_test.py

```diff
@@ -13,37 +13,98 @@
 # limitations under the License.
 """Tests server and client side compression."""
 
+from __future__ import absolute_import
+from __future__ import division
+from __future__ import print_function
+
 import unittest
 
+import contextlib
+from concurrent import futures
 import logging
+import os
+
 import grpc
+import grpc.compression
 from grpc import _grpcio_metadata
 
 from tests.unit import test_common
 from tests.unit.framework.common import test_constants
+from tests.unit import _tcp_proxy
+
+os.environ['GRPC_EXPERIMENTAL_DISABLE_FLOW_CONTROL'] = 'true'
 
 _UNARY_UNARY = '/test/UnaryUnary'
+_UNARY_STREAM = '/test/UnaryStream'
+_STREAM_UNARY = '/test/StreamUnary'
 _STREAM_STREAM = '/test/StreamStream'
 
+# Cut down on test time.
+_STREAM_LENGTH = test_constants.STREAM_LENGTH // 2
+
+
+def make_handle_unary_unary(pre_response_callback):
+
+    def _handle_unary(request, servicer_context):
+        if pre_response_callback:
+            pre_response_callback(request, servicer_context)
+        return request
+
+    return _handle_unary
+
+
+def make_handle_unary_stream(pre_response_callback):
+
+    def _handle_unary_stream(request, servicer_context):
+        if pre_response_callback:
+            pre_response_callback(request, servicer_context)
+        for _ in range(_STREAM_LENGTH):
+            yield request
+
+    return _handle_unary_stream
+
+
+def make_handle_stream_unary(pre_response_callback):
+
+    def _handle_stream_unary(request_iterator, servicer_context):
+        if pre_response_callback:
+            pre_response_callback(request_iterator, servicer_context)
+        response = None
+        for request in request_iterator:
+            if not response:
+                response = request
+        return response
+
+    return _handle_stream_unary
 
-def handle_unary(request, servicer_context):
-    servicer_context.send_initial_metadata([('grpc-internal-encoding-request',
-                                             'gzip')])
-    return request
 
+def make_handle_stream_stream(pre_response_callback):
 
-def handle_stream(request_iterator, servicer_context):
-    # TODO(issue:#6891) We should be able to remove this loop,
-    # and replace with return; yield
-    servicer_context.send_initial_metadata([('grpc-internal-encoding-request',
-                                             'gzip')])
-    for request in request_iterator:
-        yield request
+    def _handle_stream(request_iterator, servicer_context):
+        if pre_response_callback:
+            pre_response_callback(request_iterator, servicer_context)
+        # TODO(issue:#6891) We should be able to remove this loop,
+        # and replace with return; yield
+        for request in request_iterator:
+            yield request
+
+    return _handle_stream
+
+
+def set_call_compression(request_or_iterator, servicer_context):
+    del request_or_iterator
+    servicer_context.set_compression(grpc.compression.Gzip)
```

Comment:
You already have a mighty test metrics. Do you think we should test all available compression algorithms?